### PR TITLE
fix io ambiguities and add --verbose

### PR DESCRIPTION
**Pull request by mgh**
This is a fairly opinionated pull request, but there's nowhere to post issues on this repo.  Feel free to close or pick pieces from it :)

The main commit:
```
Previously, if running as a child process, imagemin would output to
stdout if the parent program tries to capture stdout.  This causes a
problem if the parent program normally writes to stdout itself, but
sometimes gets wrapped.  Further, writing multiple files back to stdout
just doesn't make sense.  Now, imagemin only pipes to stdout if it's
reading from stdin or if exactly one file is provided and no destination
directory is provided.  If a destination directory is provided, it is
always used.

Additionally, inferring directories was sometimes based on whether the
potential directory name had a dot in it, assuming that no directories
have dots in them. Now the destination directory is always assumed to be
a directory, regardless of its name.

One additional input format is allowed if no output (-o) is provided:

 $ imagemin in.png out.png

Finally, the single character command-line flags for optimization
parameters have been removed. '-p' now aliases '--plugin', and '-o' now
aliases '--output'.  This is done to prioritize the parameters that are
always valid, regardless of the plugins used.
```

### CODE REVIEW

File: cli.js [hunk 1]
```diff
@@ -4,112 +4,224 @@ var fs = require('fs');
...
-		'  $ imagemin images build',
-		'  $ imagemin foo.png > foo-optimized.png',
-		'  $ cat foo.png | imagemin > foo-optimized.png',
+		'Examples',
+		'  imagemin in.png out.png',
+		'  imagemin images/* build',
+		'  imagemin images build',
+		'  imagemin foo.png > foo-optimized.png',
+		'  imagemin < foo.png > foo-optimized.png',
+		'  imagemin -P pngquant foo.png > foo-optimized.png',
```
Comment (sindresorhus) [inline cli.js]: Happy to add more examples, but don't change the formatting.

File: cli.js [hunk 2]
```diff
@@ -4,112 +4,224 @@ var fs = require('fs');
...
 	help: [
-		'Usage',
-		'  $ imagemin <file> <directory>',
-		'  $ imagemin <directory> <output>',
-		'  $ imagemin <file> > <output>',
-		'  $ cat <file> | imagemin > <output>',
+		'Usage: imagemin [OPTION]... <in-file> <out-file>',
+		'  or   imagemin [OPTION]... <in-file> > <output>',
+		'  or   imagemin [OPTION]... [-o output-directory] <in-file>...',
+		'  or   imagemin [OPTION]... < <input> > <output>',
```
Comment (sindresorhus) [inline cli.js]: Happy to add more examples, but don't change the formatting.

File: cli.js [hunk 3]
```diff
@@ -4,112 +4,224 @@ var fs = require('fs');
...
+				console.error(files[relative].path + location
+						+ ' (' + Math.round(savings * 100) + '% smaller)');
+			}
+		});
+	}
 }
 
 if (!cli.input.length && process.stdin.isTTY) {
-	console.error([
+	exitWithMessage([
```
Comment (sindresorhus) [inline cli.js]: We can just show help here instead. `cli.showHelp();`

File: cli.js [hunk 4]
```diff
@@ -4,112 +4,224 @@ var fs = require('fs');
...
+			case 'gifsicle':
+			case 'jpegtran':
+			case 'optipng':
+				plugin = Imagemin[name](cli.flags);
+				break;
+			case 'svgo':
+				plugin = Imagemin.svgo();
+				break;
+			default:
+				// workaround pngquant bug
```
Comment (sindresorhus) [inline cli.js]: This should be fixed in the `imagemin-pngquant` module and not here.

File: cli.js [hunk 5]
```diff
@@ -4,112 +4,224 @@ var fs = require('fs');
...
+	});
+
+	if (cli.flags.verbose && srcStream !== lastStream) {
+		lastStream.on('data', function (file) {
+			if (file.contents) {
+				var relative = file.relative;
+				var savings = file.contents.length / files[relative].size;
+				var location = '';
+				if (dest) {
+					location = ' -> '
```
Comment (sindresorhus) [inline cli.js]: `->` => `→`

**Comment by sindresorhus** [comment]
Happy to land these changes if you fix the merge conflicts and resolve the inline comments.

File: cli.js [hunk 6]
```diff
@@ -4,112 +4,224 @@ var fs = require('fs');
...
+					console.error(e);
+				}
+				return undefined;
+			}
 
-		return s;
-	});
+			return s;
+		})
+		.filter(function (filename) { return filename; });
```
Comment (shinnn) [inline cli.js]: I prefer:

```js
.filter(Boolean);
```

File: cli.js [hunk 7]
```diff
@@ -4,112 +4,224 @@ var fs = require('fs');
...
+						'  npm install -g imagemin-' + name
+					]);
+				}
+				break;
+		}
+
+		imagemin.use(plugin);
+	});
+
+	var isBuffer = src instanceof Buffer;
```
Comment (shinnn) [inline cli.js]:
```js
var isBuffer = Buffer.isBuffer(src);
```

File: cli.js [hunk 8]
```diff
@@ -4,112 +4,224 @@ var fs = require('fs');
...
+	var useStdout = !process.stdout.isTTY
+			// dest was not provided?
+			&& !dest
+			// src is a buffer or single file?
+			&& (isBuffer || src.length === 1);
```
Comment (shinnn) [inline cli.js]:
```js
&& (Buffer.isBuffer(src) || src.length === 1);
```

File: cli.js (hunk 7 above)
Comment (shinnn) [inline cli.js]: And I think there is no need to assign it to a temporary variable. https://github.com/imagemin/imagemin-cli/pull/3/files#r39933387

File: cli.js [hunk 9]
```diff
@@ -4,112 +4,224 @@ var fs = require('fs');
...
-	var dest;
-
-	if (src.length > 1 && !isFile(src[src.length - 1])) {
-		dest = src[src.length - 1];
-		src.pop();
+	var files = cli.input;
+	var src;
+	var dest = cli.flags.output;
+	var destName;
+	if (!dest && files.length == 2) {
```
Comment (shinnn) [inline cli.js]: Always use [strict equal operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Comparison_Operators#Identity_strict_equality_()).

```js
if (!dest && files.length === 2) {
```

**Comment by sindresorhus** [comment]
ping @mgh

**Pull request closed without merging**